modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java

```diff
@@ -2577,21 +2568,33 @@ private void processSelectedKeys(Set<SelectionKey> keys) throws ClosedByInterrup
                     throw e;
                 }
                 catch (Exception | Error e) { // TODO IGNITE-2659.
-                    try {
-                        U.sleep(1000);
-                    }
-                    catch (IgniteInterruptedCheckedException ignore) {
-                        // No-op.
-                    }
-
-                    GridSelectorNioSessionImpl ses = attach.session();
+                    procesKeysSelectionError(e, attach);
+                }
+            }
+        }
 
-                    if (!closed)
-                        U.error(log, "Failed to process selector key [ses=" + ses + ']', e);
-                    else if (log.isDebugEnabled())
-                        log.debug("Failed to process selector key [ses=" + ses + ", err=" + e + ']');
+        /**
+         * Processes errors occured while processing selector keys.
+         *
+         * @param e Exception or error which occured.
+         * @param attach GridNioKeyAttachment.
+         */
+        private void procesKeysSelectionError(Throwable e, GridNioKeyAttachment attach) {
+            if (X.hasCause(e, Error.class)) {
```

Comment:
Can you please add a comment about this condition and sleep? I don't understand it and issues descriptions tells me nothing as well

Reply:
As I understand, this sleep was added by Alexey Goncharuk as a workaround for IGN-4019. But it shouldn't affect exceptions: holding the thread for 1 second broke other connections. All this should go away when IGNITE-2659 is fixed.